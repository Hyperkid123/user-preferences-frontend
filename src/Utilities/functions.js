import React from 'react';
import { Skeleton } from '@redhat-cloud-services/frontend-components';
import { getEmailSchema } from '../actions';
import { loaderField } from './constants';

export const getSchema = (app) => {
    return !app || !app.loaded ? loaderField : app.schema;
};

export const calculatePermissions = (permissions) => {
    return Promise.all(
        [ permissions ].flat()
        .map(({ method, args }) => insights.chrome?.visibilityFunctions?.[method]?.(...args || []))
    ).then((visibility) => visibility.every(Boolean));
};

export const calculateEmailConfig = (
    { 'email-preference': config } = { 'email-preference': {}},
    dispatch = () => {}
) => Object.entries(config).map(([ key, { permissions, url, localFile, ...rest }]) => {
    const isVisible = permissions ? calculatePermissions(permissions) : true;
    (async () => {
        const schemaVisible = await Promise.resolve(isVisible);
        if (schemaVisible) {
            if (localFile) {
                const newMapper = (await import(`../${localFile}`)).default;
                dispatch(getEmailSchema({ schema: newMapper, application: key }));
            } else {
                dispatch(getEmailSchema({ application: key, url }));
            }
        }
    })();

    return {
        [key]: {
            localFile,
            isVisible,
            url,
            ...rest
        }
    };
}).reduce((acc, curr) => ({ ...acc, ...curr }), {});

export const getSection = (key, schema = {}, storeSchema, success = () => {}) => {
    if (schema.isVisible === true) {
        return {
            label: schema?.title,
            name: key,
            fields: schema?.fields || getSchema(storeSchema)
        };
    } else {
        Promise.resolve(schema.isVisible).then(success);

        return {
            label: <Skeleton />,
            name: key,
            fields: loaderField
        };
    }
};
