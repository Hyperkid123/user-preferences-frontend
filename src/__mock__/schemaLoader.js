import { instance } from '../api';

import MockAdapter from 'axios-mock-adapter';

export const mock = new MockAdapter(instance);
