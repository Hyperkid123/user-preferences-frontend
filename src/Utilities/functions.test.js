import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { getSchema, getSection, calculatePermissions, calculateEmailConfig } from './functions';
import { loaderField } from './constants';
import { mock } from '../__mock__/schemaLoader';

describe('getSchema', () => {
    it('should return loader', () => {
        const loader = getSchema();
        expect(loader).toMatchObject(loaderField);
    });

    it('should return loader if not loaded', () => {
        const loader = getSchema({ loaded: false });
        expect(loader).toMatchObject(loaderField);
    });

    it('should return schema', () => {
        const schema = { fields: []};
        const loaded = getSchema({ loaded: true, schema });
        expect(loaded).toMatchObject(schema);
    });
});

describe('getSection', () => {
    it('should return visible schema', () => {
        const section = getSection('first', { isVisible: true });
        expect(section).toMatchObject({ name: 'first', fields: loaderField });
    });

    it('should return visible schema from store', () => {
        const schema = { fields: []};
        const section = getSection('first', { isVisible: true, title: 'Some title' }, { loaded: true, schema });
        expect(section).toMatchObject({
            label: 'Some title',
            name: 'first',
            fields: schema
        });
    });

    it('should return loader', () => {
        const section = getSection();
        const wrapper = mount(section.label);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(section).toMatchObject({ fields: loaderField });
    });

    it('should call success function', (done) => {
        const success = jest.fn();
        const section = getSection('first', { isVisible: false }, {}, success);
        expect(section).toMatchObject({
            name: 'first',
            fields: loaderField
        });
        setTimeout(() => {
            expect(success).toHaveBeenCalled();
            done();
        });
    });
});

describe('calculatePermissions', () => {
    it('should check visibility of one function', async () => {
        const isVisible = await calculatePermissions({
            method: 'something'
        });
        expect(isVisible).toBe(false);
    });

    it('should check visibility of array of functions', async () => {
        const isVisible = await calculatePermissions([{
            method: 'something'
        }, {
            method: 'something',
            args: [ true ]
        }]);
        expect(isVisible).toBe(false);
    });

    it('should calculate visible for one function', async () => {
        const isVisible = await calculatePermissions([{
            method: 'something',
            args: [ true ]
        }]);
        expect(isVisible).toBe(true);
    });
});

describe('calculateEmailConfig', () => {
    it('should not throw error', async () => {
        const result = await calculateEmailConfig();
        expect(result).toMatchObject({});
    });

    it('should calculate schema with permissions - false', async () => {
        mock.onGet('/api*').reply(200, {});
        const result = calculateEmailConfig({
            'email-preference': {
                test: {
                    permissions: { method: 'something' }
                }
            }
        });
        expect(await result.test.isVisible).toBe(false);
    });

    it('should calculate schema with permissions - true', async () => {
        mock.onGet('/api/test/v1/user-config/email-preference').reply(200, {});
        const result = calculateEmailConfig({
            'email-preference': {
                test: {
                    permissions: { method: 'something', args: [ true ]}
                }
            }
        });
        expect(await result.test.isVisible).toBe(true);
    });

    it('should request localFile', (done) => {
        mock.onGet('/api/test/v1/user-config/email-preference').reply(200, {});
        const dispatch = jest.fn();
        calculateEmailConfig({
            'email-preference': {
                test: {}
            }
        }, dispatch);
        setImmediate(() => {
            expect(dispatch).toHaveBeenCalled();
            expect(dispatch.mock.calls[0][0]).toMatchObject({
                meta: {
                    appName: 'test'
                }
            });
            done();
        });
    });

    it('should request localFile', (done) => {
        const dispatch = jest.fn();
        calculateEmailConfig({
            'email-preference': {
                test: {
                    localFile: 'data/general.json'
                }
            }
        }, dispatch);
        setImmediate(() => {
            expect(dispatch).toHaveBeenCalled();
            expect(dispatch.mock.calls[0][0]).toMatchObject({
                payload: {},
                meta: {
                    appName: 'test'
                }
            });
            done();
        });
    });
});
