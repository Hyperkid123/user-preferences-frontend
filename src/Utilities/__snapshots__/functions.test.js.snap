// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSection should return loader 1`] = `
<Skeleton
  isDark={false}
  size="md"
>
  <div
    className="ins-c-skeleton ins-c-skeleton__md"
  >
    Â 
  </div>
</Skeleton>
`;
